Azure Monitor Exporter - Adopt allocation less ForEach from OpenTelemetry

rajkumar-rangaraj opened the pull request:
Addressing https://github.com/Azure/azure-sdk-for-net/pull/15533#discussion_r498453868 

Adopting https://github.com/open-telemetry/opentelemetry-dotnet/pull/1236

rajkumar-rangaraj committed:
Otel ForEach

reyang commented on sdk/monitor/Microsoft.OpenTelemetry.Exporter.AzureMonitor/src/EnumerationHelper.cs:
```diff
@@ -0,0 +1,140 @@
+// <auto-generated>
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
This file was copied over, which has license/legal issue.

reyang commented:
What's the perf gain? It would be helpful to see the benchmark diff.

rajkumar-rangaraj committed:
Add Benchmark

rajkumar-rangaraj committed:
Complete Benchmarks

rajkumar-rangaraj commented:
@reyang 

Usage of `PooledList' reduces the data allocation completely with better execution time. This PR has two changes

1. Using `PooledList` to store `PartB` and `PartC` tags instead of `Dictionary`
2. Customized `GetValue` method which is faster than default dictionary implementation and Otel's `Activity.GetTagValue`

``` ini

BenchmarkDotNet=v0.11.5, OS=Windows 10.0.19042
Intel Core i7-8650U CPU 1.90GHz (Kaby Lake R), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=3.1.404
  [Host]     : .NET Core 2.1.23 (CoreCLR 4.6.29321.03, CoreFX 4.6.29321.01), 64bit RyuJIT
  DefaultJob : .NET Core 2.1.23 (CoreCLR 4.6.29321.03, CoreFX 4.6.29321.01), 64bit RyuJIT

```
* Enumeration of complete span attributes

|                           Method |         Mean |      Error |     StdDev |  Gen 0 |  Gen 1 | Gen 2 | Allocated |
|--------------------------------- |-------------:|-----------:|-----------:|-------:|-------:|------:|----------:|
|      Enumerate_PooledList_NoItem |     4.830 ns |  0.1647 ns |  0.4752 ns |      - |      - |     - |         - |
|      Enumerate_PooledList_Part_B |   311.643 ns | 11.0499 ns | 30.6193 ns |      - |      - |     - |         - |
|      Enumerate_PooledList_Part_C | 1,102.084 ns | 21.9438 ns | 43.3148 ns | 0.0534 | 0.0134 |     - |     344 B |
| Enumerate_PooledList_PartB_And_C |   288.731 ns |  5.6775 ns | 10.9385 ns |      - |      - |     - |         - |
|      Enumerate_TagObjects_NoItem |   278.346 ns | 11.8114 ns | 34.4546 ns | 0.0515 |      - |     - |     216 B |
|       Enumerate_TagObjects_PartB |   700.596 ns | 12.8913 ns | 12.0586 ns | 0.1421 |      - |     - |     600 B |
|       Enumerate_TagObjects_PartC | 1,383.998 ns | 27.3425 ns | 48.6012 ns | 0.2403 |      - |     - |    1008 B |
| Enumerate_TagObjects_PartB_And_C |   634.168 ns | 12.4498 ns | 23.3837 ns | 0.1154 |      - |     - |     488 B |

* Get a specific value from span attribute

|                          Method |        Mean |     Error |    StdDev | Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------------------------- |------------:|----------:|----------:|------:|------:|------:|----------:|
|      GetTagValueEmptyTagObjects |   3.9359 ns | 0.1088 ns | 0.1336 ns |     - |     - |     - |         - |
|   GetTagValueNonemptyTagObjects |  19.1331 ns | 0.2951 ns | 0.2616 ns |     - |     - |     - |         - |
|      GetTagValueEmptyPooledList |   0.7019 ns | 0.0482 ns | 0.0610 ns |     - |     - |     - |         - |
|   GetTagValueNonemptyPooledList |  19.7032 ns | 0.3117 ns | 0.2603 ns |     - |     - |     - |         - |
|    GetTagValueEmptyActivityTags |   5.1585 ns | 0.1886 ns | 0.2453 ns |     - |     - |     - |         - |
| GetTagValueNonemptyActivityTags | 100.3957 ns | 3.3325 ns | 2.7828 ns |     - |     - |     - |         - |


* Removed `EnumerationHelper.cs` which was from OpenTelemetry repo.
* Still `PooledList.cs` is an implementation from OpenTelemetry repo. Need your guidance to check if we could take copy of this file in our exporter.

reyang commented:
> Adopting [open-telemetry/opentelemetry-dotnet#1236](https://github.com/open-telemetry/opentelemetry-dotnet/pull/1236)
> 
> `PooledList.cs` and `EnumerationHelper.cs` are from Otel repo. This is marked as internal in Otel hence could not take direct reference.

@cijothomas @rajkumar-rangaraj would it be possible to access these via reflection + cache, instead of copying the file over? I'm a bit concerned about the license issue.

rajkumar-rangaraj committed:
Introduce AzMonList and remove all OTel source reference.

rajkumar-rangaraj commented:
Removed all source from OpenTelemetry repo. Taken guidance from [Minimize large object allocations](https://docs.microsoft.com/en-us/aspnet/core/performance/performance-best-practices?view=aspnetcore-5.0#minimize-large-object-allocations), learning from `PooledList.cs` and introduced a new non-generic struct named `AzMonList`. We could achieve a similar functionality as above with new implementation. Below is the benchmark on new code.

**Enumerate Tags**

|                           Method |         Mean |      Error |     StdDev |       Median |  Gen 0 |  Gen 1 | Gen 2 | Allocated |
|--------------------------------- |-------------:|-----------:|-----------:|-------------:|-------:|-------:|------:|----------:|
|       Enumerate_AzMonList_NoItem |     4.442 ns |  0.1302 ns |  0.2140 ns |     4.433 ns |      - |      - |     - |         - |
|       Enumerate_AzMonList_Part_B |   295.398 ns |  8.8381 ns | 25.2156 ns |   285.778 ns |      - |      - |     - |         - |
|       Enumerate_AzMonList_Part_C | 1,048.600 ns | 20.8591 ns | 18.4911 ns | 1,045.783 ns | 0.0534 | 0.0134 |     - |     344 B |
|  Enumerate_AzMonList_PartB_And_C |   277.047 ns |  5.4749 ns |  6.9240 ns |   277.856 ns |      - |      - |     - |         - |
|      Enumerate_TagObjects_NoItem |   221.803 ns |  4.1953 ns |  3.9242 ns |   222.323 ns | 0.0515 |      - |     - |     216 B |
|       Enumerate_TagObjects_PartB |   708.207 ns | 13.8387 ns | 16.4740 ns |   708.811 ns | 0.1421 |      - |     - |     600 B |
|       Enumerate_TagObjects_PartC | 1,378.092 ns | 26.5406 ns | 31.5947 ns | 1,381.193 ns | 0.2403 |      - |     - |    1008 B |
| Enumerate_TagObjects_PartB_And_C |   626.087 ns | 12.0612 ns | 11.8457 ns |   624.068 ns | 0.1154 |      - |     - |     488 B |

**GetValue**

|                          Method |        Mean |     Error |    StdDev | Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------------------------- |------------:|----------:|----------:|------:|------:|------:|----------:|
|      GetTagValueEmptyTagObjects |   4.3709 ns | 0.1314 ns | 0.3642 ns |     - |     - |     - |         - |
|   GetTagValueNonemptyTagObjects |  21.3269 ns | 0.5158 ns | 1.4034 ns |     - |     - |     - |         - |
|       GetTagValueEmptyAzMonList |   0.6705 ns | 0.0676 ns | 0.1863 ns |     - |     - |     - |         - |
|    GetTagValueNonemptyAzMonList |  23.9153 ns | 0.6331 ns | 1.8064 ns |     - |     - |     - |         - |
|    GetTagValueEmptyActivityTags |   5.5623 ns | 0.1689 ns | 0.4596 ns |     - |     - |     - |         - |
| GetTagValueNonemptyActivityTags | 101.1347 ns | 2.0652 ns | 5.5834 ns |     - |     - |     - |         - |

reyang commented on sdk/monitor/Microsoft.OpenTelemetry.Exporter.AzureMonitor/src/AzureMonitorConverter.cs:
```diff
@@ -121,16 +169,23 @@ internal static void InitRoleInfo(Activity activity)
-            var activityType = activity.TagObjects.ToAzureMonitorTags(out var partBTags, out var PartCTags);
+            var monitorTags = new TagEnumerationState
+            {
+                PartBTags =AzMonList.Initialize(),
```
```suggestion
                PartBTags = AzMonList.Initialize(),
```

reyang reviewed (approved):
LGTM.

rajkumar-rangaraj committed:
Add space before AzMonList.Initialize
